paddle/framework/tensor.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include "paddle/framework/ddim.h"
+#include "paddle/platform/assert.h"
+#include "paddle/platform/place.h"
+
+namespace paddle {
+namespace framework {
+
+class Tensor {
+  using paddle::platform::Place;
+  using paddle::platform::get_place;
+
+ public:
+  explicit Tensor(DDim dims) : dims_(dims), place_(get_place()) {}
+  explicit Tensor(DDim dims, Place place) : dims_(dims), place_(place) {}
+
+  Tensor& operator=(const Tensor& src) = delete;
+
+  template <typename T>
+  const T* Data() const;
+
+  bool NeedReset();
+
+  // must be POD types
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData();
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData(const DDim& dims);
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData(const DDim& dims, const Place& place);
+
+  int Rank() const;
+
+  int Numel() const;
+
+  void Resize(const DDim& dims);
+
+  void Reshape(const DDim& dims);
+
+  void ShareData(const Tensor& src);
+
+  template <typename T>
+  void CopyFrom(const Tensor& src);
+
+  const std::shared_ptr<Placeholder>& Holder() const;
+
+  const DDim& Dims() const;
+
+  const paddle::platform::Place& Place() const;
+
+  template <typename T>
+  bool IsType() const;
+
+  // Placeholder hides type T, so it doesn't appear as a template
+  struct Placeholder {
+    virtual ~Placeholder() {}
+    virtual std::type_info TypeInfo() const = 0;
+    virtual void* Ptr() const = 0;
+    virtual Place Place() const = 0;
+    virtual size_t Size() const = 0;
+  };
+
+ private:
+  template <typename T>
+  struct PlaceholderImpl : public Placeholder {
+    PlaceholderImpl(Place place, size_t size)
+        : ptr_(paddle::memory::Alloc(place, size),
+               paddle::memory::Deleter(place)),
+          place_(place),
+          size_(size) {}
+
+    virtual std::type_info TypeInfo() const { return typeid(T); }
+    virtual void* Ptr() const { return static_cast<void*>(ptr_.get()); }
```

Comment:
Maybe Ptr() should not be added in interface, it should return T* like

```cpp
struct Placeholder {
  virtual ~Placeholder() {}  // for rtti
};

template <typename T>
struct PlaceholderImpl : public Placeholder {
  T* Ptr() const { return ptr_.get(); }
};

template <typename T>
const T* Data() const {
  auto holder = std::dynamic_pointer_cast<PlaceholderImpl<T>>(holder_);
  ASSERT(holder != nullptr);
  return holder->Ptr();
}
```
Using `static_cast` everywhere make compiler cannot check type for us.